server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -162,6 +169,18 @@ protected SearchHit nextDoc(int doc) throws IOException {
                 if (context.isCancelled()) {
                     throw new TaskCancelledException("cancelled");
                 }
+                ++processedDocs;
+                // indicates if we should check with the parent circuit breaker (the real memory breaker) if we have enough memory to
+                // continue the fetch phase.
+                // we do so whenever one of the following is true:
+                // 1. we have accumulated at least the size of the memory accounting buffer
+                // 2. we have reached the last document in the leaf
+                boolean enoughBytesOrLastDocInLeaf = (accumulatedBytesInLeaf >= context.memAccountingBufferSize())
+                    || (processedDocs == docsInLeaf);
+                if (enoughBytesOrLastDocInLeaf) {
+                    context.circuitBreaker().addEstimateBytesAndMaybeBreak(0, "fetch source");
```

Comment:
I wonder: should we just make this call an implementation detail of the context?
i.e. just have a `void checkBreaker()` call like we do in aggs? Makes it a little easier to remove the real memory breaker surgically down the line :) That's why the API is the way it is in aggs as well I think.

Reply:
I thought about it initially. The condition we have here is more specific to fetch phase and how we accumulate / segment (vs the aggs where it's easy to generically bake in the `checkBreaker` method) 

We could still do something similar here I think, if we have a `checkBreaker` that takes a `BooleanPredicate` so callers can indicate when to hit the breaker?

Reply:
Made this work https://github.com/elastic/elasticsearch/pull/121920/commits/06dcb68127721dd9333e2b1429355439e13e1b40